activestorage/app/models/active_storage/blob.rb

```diff
@@ -36,10 +36,24 @@ class ActiveStorage::Blob < ActiveRecord::Base
 
   scope :unattached, -> { left_joins(:attachments).where(ActiveStorage::Attachment.table_name => { blob_id: nil }) }
 
+  after_initialize do
+    self.service_name ||= Rails.configuration.active_storage.service
+  end
+
   before_destroy(prepend: true) do
     raise ActiveRecord::InvalidForeignKey if attachments.exists?
   end
 
+  validates :service_name, presence: true
+
+  validate do
```

Comment:
I wonder what the overhead for unchanged service names would be? It might be worth validating only if the attribute changed.